src/storage/src/render/upsert/types.rs

```diff
@@ -7,85 +7,478 @@
-    state: HashMap<UpsertKey, UpsertValue>,
+    state: HashMap<UpsertKey, StateValue>,
+
+    // Bincode options and buffer used
+    // in `merge_snapshot_chunk`.
+    bincode_opts: BincodeOpts,
+    bincode_buffer: Vec<u8>,
 }
 
 impl Default for InMemoryHashMap {
     fn default() -> Self {
         Self {
             state: HashMap::new(),
+            bincode_opts: upsert_bincode_opts(),
+            bincode_buffer: Vec::new(),
         }
     }
 }
 
 #[async_trait::async_trait(?Send)]
 impl UpsertState for InMemoryHashMap {
-    async fn multi_put<P>(&mut self, puts: P) -> Result<u64, anyhow::Error>
+    const SNAPSHOT_BATCH_SIZE: usize = 1;
+    async fn merge_snapshot_chunk<P>(&mut self, puts: P) -> Result<MergeStats, anyhow::Error>
+    where
+        P: IntoIterator<Item = (UpsertKey, UpsertValue, mz_repr::Diff)>,
+    {
+        let mut stats = MergeStats::default();
+        for (key, value, diff) in puts {
+            stats.updates += 1;
+
+            let entry = self.state.entry(key);
+            let current_size: i64 = match &entry {
+                hash_map::Entry::Occupied(o) => o
+                    .get()
+                    .memory_size()
+                    .try_into()
+                    .expect("less than i64 size"),
+                hash_map::Entry::Vacant(_) => {
+                    stats.values_diff += 1;
+                    0
+                }
+            };
+            let current_value = entry.or_insert_with(Default::default);
+
+            if current_value.merge_update(value, diff, self.bincode_opts, &mut self.bincode_buffer)
+            {
+                stats.values_diff -= 1;
+                stats.size_diff -= current_size;
+                self.state.remove(&key);
+            } else {
+                stats.size_diff -= current_size;
+                let current_size: i64 = current_value
+                    .memory_size()
+                    .try_into()
+                    .expect("less than i64 size");
+                stats.size_diff += current_size
+            }
+        }
+
+        Ok(stats)
+    }
+
+    async fn multi_put<P>(&mut self, puts: P) -> Result<PutStats, anyhow::Error>
     where
-        P: IntoIterator<Item = (UpsertKey, Option<UpsertValue>)>,
+        P: IntoIterator<Item = (UpsertKey, PutValue)>,
     {
-        let mut size = 0;
-        for (key, value) in puts {
-            match value {
+        let mut stats = PutStats::default();
+        for (key, p_value) in puts {
+            stats.processed_puts += 1;
+            match p_value.value {
                 Some(value) => {
-                    self.state.insert(key, value);
+                    let decoded_value = StateValue::Decoded(value);
+                    let size: i64 = decoded_value
+                        .memory_size()
+                        .try_into()
+                        .expect("less than i64 size");
+                    match p_value.previous_persisted_size {
```

Comment:
No, because the `envelope_state_bytes` will be seeded with the original (larger) size. `multi_put` is allowed to return negative numbers of the size has gone down (cause updates are smaller, or lots of deletes)

Reply:
Say, we have only one value in the state and say it's size when `Snapshotting` is 1 and when `Decoded`, it's 2.

Originally `envelope_state_bytes` will have 1. If the record is now removed, we'll be subtracting 2 since it's in `Decoded` form. Overall `envelope_state_bytes` is now weirdly -1.